add parallel phpmd processing and polish

ablyler opened the pull request:
- use forking to process files in parallel (need for large code bases)
- add support for specifing rulesets in the config

Note: I switched to from Alpine to Ubuntu for now because Alpine doesn't have php-pcntl support, which was needed for forking. I submitted a [patch](http://lists.alpinelinux.org/alpine-aports/0853.html) to Alpine to get php-pcntl support added.

ablyler committed:
add parallel phpmd processing and polish

- use forking to process files in parallel (need for large code bases)
- add support for specifing rulesets in the config

brynary commented on engine.php:
```diff
@@ -1,70 +1,42 @@
-  return $result;
+foreach ($server->get_all_results() as $result_file) {
+    echo file_get_contents($result_file);
+    unlink($result_file);
```
Where do result files get stored? (We'll need to ensure it's within `/tmp`)

brynary replied:
Just saw the reference to `sys_get_temp_dir`, so this should be OK.

ablyler replied:
You got it.

brynary commented on engine.php:
```diff
@@ -1,70 +1,42 @@
+// setup forking daemon
+$server = new \fork_daemon();
+$server->max_children_set(20);
+$server->max_work_per_child_set(50);
```
Can you please help me understand what these two values do? (20 and 50)

ablyler replied:
Sure, `max_children_set` with the value of 20 means that the script with create up to 20 forks / processes, each processing up to 50 pieces of work (`max_work_per_child_set`).  The pieces of work in this case are the PHP files to scan.

brynary commented on Dockerfile:
```diff
@@ -1,17 +1,24 @@
-FROM alpine:edge
+FROM phusion/baseimage
```
We should probably pin this to a specific tag of `phusion/baseimage` for good measure. Looks like 0.9.17 is latest right now.

ablyler replied:
Good idea, will do.

brynary commented on README.md:
```diff
@@ -1,4 +1,36 @@
+Format the values for these config options per the [PHP_CodeSniffer documentation](https://github.com/squizlabs/PHP_CodeSniffer).
+
+* file_extensions - This is where you can configure the file extensions for the files that you want PHP_CodeSniffer to analyze.
+* rulesets - This is the list of rulesets that you want PHPMD to use while analyzing your files.
```
There are some references to PHP_CodeSniffer here that probably shouldn't be here.

ablyler replied:
Good old copy and paste, good catch, will fix.

brynary commented on README.md:
```diff
@@ -1,4 +1,36 @@
+      phpmd:
+        enabled: true
+        config:
+          file_extensions: "php"
```
We should probably set up the `file_extensions` key so that it can work with a single value like this, or an array:

``` yaml
file_extensions:
  - php
  - php5
```

ablyler replied:
Yeah, I believe I just copied this from the codeclimate-phpcodesniffer repo.  Maybe we should double check and then address it separately?

brynary replied:
Agreed, separate sounds good!

brynary commented on README.md:
```diff
@@ -1,4 +1,36 @@
+        enabled: true
+        config:
+          file_extensions: "php"
+          rulesets: "unusedcode"
```
Similarly, my gut is we should allow either a single ruleset like this (`ruleset: unusedcode`) as a one-liner, or:

``` yaml
rulesets:
  - unusedcode
  - design
  - naming
```

The engine can implode the array.

ablyler replied:
Yeah, I believe I just copied this from the codeclimate-phpcodesniffer repo.  Maybe we should double check and then address it separately?

brynary commented on Runner.php:
```diff
@@ -0,0 +1,77 @@
+use PHPMD\Writer\StreamWriter;
+use PHPMD\Renderer\JSONRenderer;
+
+class Runner
```
Nice extraction here.

ablyler replied:
:+1:

brynary commented on Runner.php:
```diff
@@ -0,0 +1,77 @@
+    private $config;
+    private $server;
+
+    public function setConfig($config)
```
Since a `Runner` always needs the `$config` and `$server` members to work properly, thoughts on making them part of the constructor?

ablyler replied:
Good idea, will do.

brynary commented:
Hi @ablyler -- Thanks for this. I think this is going to be a big improvement. I added some line notes for questions/suggestions, but this seems really close.

ablyler committed:
Pin Docker to using 0.9.17 of phusion/baseimage

ablyler committed:
Fix references to PHPCS in README

ablyler committed:
Add config and server to constructor of Runner

ablyler committed:
Revert to using alpine linux as the base

- we can use it now that it has php-pcntl support

ablyler commented:
@brynary everything should be all fixed up. also just changed the base back to alpine linux, now that they accepted my php-pcntl addition. :-)

brynary commented:
@ablyler looks great. Ready to merge, just one more thing... 20 forks processing files simultaneously is probably too many (which would cause a slowdown due to unnecessary CPU switching).

For our CodeClimate.com system, we'll allocate 2 CPUs to this, so we should probably run with a concurrency of about 2 or 3. (We can allow this to be overridden with an environment variable for local usage and on-premise deployments with different resource constraints as well, in the future.)

ablyler committed:
Change max number of children from 20 to 3

- did some performance testing after @brynary comment in GitHub and found that 3 performs best on my large code bases

ablyler commented:
@brynary good call, did some testing, 3 appears to be a good setting.

brynary commented:
Merged. Thanks for all the help on this, @ablyler!
